src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,415 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
+{
+    if (!gdata->assertOn) {
+        return;
+    }
+    if (gdata->vmDead) {
+        return; // This assert is not reliable if the VM is exiting
+    }
+    if (!isSameObject(env, thread, current_thread)) {
+        tty_message("ERROR: Threads are not the same: \"%s\" \"%s\"",
+                    getThreadName(thread), getThreadName(current_thread));
+        dumpRawMonitors();
+        JDI_ASSERT(JNI_FALSE);
+    }
+}
+
+static void
+verifyMonitorRank(JNIEnv *env, DebugRawMonitorRank rank, jthread thread)
+{
+    // We must hold the dbgRawMonitor when calling verifyMonitorRank()
+
+    // Iterate over all the monitors and makes sure we don't already hold one that
+    // has a higher rank than the monitor we are about to enter.
+    DebugRawMonitorRank i;
+    DebugRawMonitorRank firstRank =
+        (rank < FIRST_LEAF_DEBUG_RAW_MONITOR ? rank : FIRST_LEAF_DEBUG_RAW_MONITOR);
+    for (i = firstRank; i < NUM_DEBUG_RAW_MONITORS; i++) {
+        DebugRawMonitor* dbg_monitor = &dbg_monitors[i];
+        if (dbg_monitor->monitor == NULL) {
+            continue; // ignore uninitialzed monitors
+        }
+        JDI_ASSERT(i == dbg_monitor->rank);
+        if (dbg_monitor->ownerThread == NULL) {
+            continue; // ignore unowned monitors
+        }
+        if (!isSameObject(env, dbg_monitor->ownerThread, thread)) {
+            continue; // ignore monitors not owned by this thread
+        }
+
+        /*
+         * dbg_monitor is a monitor that the thread currently holds. We need to
+         * do two verificatons of its rank:
+         *   1. The held monitor's rank is not higher than the rank of the monitor
+         *      we are trying to enter.
+         *   2. The held monitor is not a leaf monitor, because we should not be entering
+         *      any additional monitors while holding a leaf monitor.
+         */
+        if (dbg_monitor->rank > rank || dbg_monitor->rank >= FIRST_LEAF_DEBUG_RAW_MONITOR ) {
+          // At this point we have found a rank order violation.
+          if (dbg_monitor->rank == popFrameEventLock_Rank && rank == popFrameProceedLock_Rank) {
+              // For reasons too complex to explain here, this is one exception that is safe so
+              // we allow it. See the long comment for these locks in threadControl.c.
+              continue;
+          }
+          char* threadName = getThreadName(thread);
+          if (dbg_monitor->rank > rank) {
+              tty_message("DebugRawMonitor rank failure: (%s: %d > %d) for thread (%s)",
+                          dbg_monitor->name, dbg_monitor->rank, rank, threadName);
+          } else {
+              tty_message("DebugRawMonitor rank failure: (%s is a held leaf monitor) for thread (%s)",
+                          dbg_monitor->name, threadName);
+          }
+          jvmtiDeallocate(threadName);
+          dumpRawMonitors();
+          JDI_ASSERT(JNI_FALSE);
+        }
+    }
+}
```

Comment:
Nit: I'd suggest to consider two separate loops instead of one, something like below:

```
static void
assertOrderFailure(jthread thread,  DebugRawMonitorRank rank, DebugRawMonitor* dbg_monitor, char* msg)
{
    char* threadName = getThreadName(thread);
     tty_message("DebugRawMonitor %s failure: (%s: %d > %d) for thread (%s)",
                 msg, dbg_monitor->name, dbg_monitor->rank, rank, threadName);
     jvmtiDeallocate(threadName);
     JDI_ASSERT(JNI_FALSE);
}

static void
verifyMonitorRank(JNIEnv *env, DebugRawMonitorRank rank, jthread thread)
{
    DebugRawMonitorRank i;

    // 1. Verify the held monitor's rank is lower than the rank of the monitor we are trying to enter.
    for (i = rank + 1; i < FIRST_LEAF_DEBUG_RAW_MONITOR; i++) {
        DebugRawMonitor* dbg_monitor = &dbg_monitors[i];
        if (dbg_monitor->ownerThread != NULL &&
            isSameObject(env, dbg_monitor->ownerThread, thread)) {
            // the lower ranked monitor is owned by this thread
            assertOrderFailure(thread, rank, dbg_monitor, "RANK ORDER");
        }
    }
    // 2. Verify there are no other leaf monitors owned but this thread.
    for (i = FIRST_LEAF_DEBUG_RAW_MONITOR; i < NUM_DEBUG_RAW_MONITORS; i++) {
        DebugRawMonitor* dbg_monitor = &dbg_monitors[i];
        if (i != rank && isSameObject(env, dbg_monitor->ownerThread, thread)) {
            // the leaf ranked monitor is already owned by this thread
            assertOrderFailure(thread, rank, dbg_monitor, "LEAF RANK");
        }
    }
```
I hope, this can be tweaked to adopt the `popFrame` locks correction.

Reply:
The special popFrame check needs to go in the first loop only, so it shouldn't be a problem or add any complexity that we don't already have.

One things to resolve is if we enter a regular monitor while holding a leaf monitor, is this a "rank" failure or "leaf" failure. Currently in the code it is a "rank" failure. Your change would make it a "leaf" failure.

I'm not sure why you added the "i != rank" check with the leaf check. We should never be re-entering a leaf monitor. The same "dbg_monitor->ownerThread != NULL" check as in the first loop should be used.

assertOrderFailure() won't be able to print as descriptive of a message as what I currently have.

Reply:
> The special popFrame check needs to go in the first loop only, so it shouldn't be a problem or add any complexity that we don't already have.

Sounds good.

> One things to resolve is if we enter a regular monitor while holding a leaf monitor, is this a "rank" failure or "leaf" failure. Currently in the code it is a "rank" failure. Your change would make it a "leaf" failure.

A "leaf" failure is more specific than a "rank order" failure, so it is better to report it first. Each "leaf" failure is also  a "rank order" failure (AFAICS).

> I'm not sure why you added the "i != rank" check with the leaf check. We should never be re-entering a leaf monitor. The same "dbg_monitor->ownerThread != NULL" check as in the first loop should be used.

You are right. This check is not needed and has to be removed. I was thinking a leaf monitor can be grabbed recursively.

Reply:
> A "leaf" failure is more specific than a "rank order" failure, so it is better to report it first. Each "leaf" failure is also a "rank order" failure (AFAICS).

It's not just a matter of which is reported first. Even if you swap the order of your two loops you get the same result. The problem is the "rank" loop does not check if any of the leaf monitors are already held. I think to fix that it would have to iterate up to NUM_DEBUG_RAW_MONITORS, which means there is overlap with the range that the "leaf" loop iterators over.

Reply:
In fact, I do not understand why reporting the "rank order" violation is important what "leaf rank order" is violated.
I feel that I'm missing something. Let me think on it a little bit.

Reply:
If the following are all true 
- you are entering is a leaf monitor
- the current thread has already entered a leaf monitor
- the rank of the monitor you are entering is lower than the rank of the monitor already held

Then you have both a rank order violation and a violation for entering a leaf monitor when you have already entered a leaf monitor. My current implementation will complain about the rank violation (although can easily be made to instead complain about the leaf violation or complain about both). Yours will complain about the leaf violation.  To make yours instead complain about the rank violation, the first loop needs to also check all the leaf monitors. Since the 2nd loop also checks all the leaf monitors (but for a different reason), this means iterating over the leaf monitors twice. Complaining about both violations would also require iterating over the leaf monitors twice.

Reply:
Okay, thanks.
> Then you have both a rank order violation and a violation for entering a leaf monitor when you have already entered a leaf monitor. 

I kind of disagree with the second part of this statement.
My understanding is that the second loop does a little bit different check:
  "Violation for entering ANY monitor when you have already entered a leaf monitor."
Please, note that `rank` can be any rank including those with condition:
```
   rank < FIRST_LEAF_DEBUG_RAW_MONITOR
```

Reply:
Yes, the 2nd loop is a different check. That's why I said it also checks all the leaf monitors "but for a different reason".  Your two loops do not flag a rank violation if both monitors are leafs, even if grabbed in the wrong order. It only flags the leaf violation. Your two checks will always catch any violation, it just a matter of whether my example (which is both a leaf and a rank violation) is flagged as a leaf violation or a rank violation (or even both could be indicated if we choose). Yours flags it as a leaf violation. My code flags it as a rank violation. Mine could flag both violations without any additional iterations. Your would need to iterate over the leaf monitors twice to detect if there is both a rank and a leaf violation.

Reply:
Okay, thanks. Please, let me list the variants with some analysis.
We have 3 variants:
1.  Both monitors are leaf: Entering any monitor while a leaf monitor has been entered is a violation (`LEAF RULE` violation). The order does not matter. Any order is unacceptable.

2. A non-leaf monitor is being entered while a leaf monitor has been entered: This is also a `LEAF RULE` violation. This violation is at the same time always a "RANK ORDER VIOLATION" for the non-leaf monitor. My view is there is no need to report this "RANK ORDER VIOLATION" as it always coexists with the `LEAF RULE` violation for non-leaf monitors. But if you insist on reporting it additionally then it is not a problem to check and report it in the `assertOrderFailure()` function. It has all needed info available for it.

3. A non-leaf monitor is being entered while a non-leaf monitor has been entered: It is a case+report for "RANK ORDER VIOLATION".

There is one more variant:
4. A leaf monitor is being entered while a non-leaf monitor has been entered: It is never a violation, so this variant is excluded from the list above.

Reply:
But 2 and 1 are very different. You can call them both leaf violations, but they are leaf violations for very different reasons, and 2 is more akin to a rank violation than a leaf violation.

I'm reversing the ranks and reworking the loop a bit (both the comments and how the errors are reported). I'll try to post later tonight after testing is done.

Reply:
I've given some more thought to this "leaf monitor" concept and decided it's not worth having around. The only thing that results in branding a monitor as a leaf monitor is the fact that no additional monitors are locked while the leaf monitor is held. This is not actually a requirement, but simply an observation made for these monitors. If any additional locking was introduced while holding a leaf monitor, that is not necessarily a bad thing. The list of leaf monitors is simply what was left over at the bottom of the rankings after all monitors that needed a higher rank were given one (and this was the result of an iterative process of running tests and seeing what asserts were triggered because of misranking some monitors). This list started out long, but after fixing all the rank issues we ended up with just 4.